src/Symfony/Component/Workflow/Marking.php

```diff
@@ -27,23 +27,56 @@ class Marking
     public function __construct(array $representation = [])
     {
         foreach ($representation as $place => $nbToken) {
-            $this->mark($place);
+            $this->mark($place, $nbToken);
         }
     }
 
-    public function mark(string $place): void
+    /**
+     * @param int $nbToken
+     */
+    public function mark(string $place /* , int $nbToken = 1 */): void
     {
-        $this->places[$place] = 1;
+        $nbToken = 1 < \func_num_args() ? func_get_arg(1) : 1;
+
+        if ($nbToken < 1) {
+            throw new \InvalidArgumentException(sprintf('The number of tokens must be greater than 0, "%s" given.', $nbToken));
+        }
+
+        if (!\array_key_exists($place, $this->places)) {
+            $this->places[$place] = 0;
+        }
+        $this->places[$place] += $nbToken;
     }
 
-    public function unmark(string $place): void
+    /**
+     * @param int $nbToken
+     */
+    public function unmark(string $place /* , int $nbToken = 1 */): void
     {
-        unset($this->places[$place]);
+        $nbToken = 1 < \func_num_args() ? func_get_arg(1) : 1;
+
+        if ($nbToken < 1) {
+            throw new \InvalidArgumentException(sprintf('The number of tokens must be greater than 0, "%s" given.', $nbToken));
+        }
+
+        if (!$this->has($place)) {
+            throw new \InvalidArgumentException(sprintf('The place "%s" is not marked.', $place));
+        }
+
+        $this->places[$place] -= $nbToken;
+
+        if (0 > $this->places[$place]) {
+            throw new \InvalidArgumentException(sprintf('The place "%s" could not contain a negative token number.', $place));
```

Comment:
I suppose this check should happen before the subtraction so that we don't mutate the object and then trow

Reply:
Very good catch, thanks. Fixed